include/sampleflow/producers/delayed_rejection_mh.h

```diff
@@ -0,0 +1,172 @@
+         *   sample is available.
+         */
+        void
+        sample (const OutputType &starting_point,
+                const std::function<double (const OutputType &)> &log_likelihood,
+                const std::function<std::pair<OutputType,double> (const OutputType &, const std::vector<OutputType>)> &perturb,
+                const unsigned int max_delays,
+                const types::sample_index n_samples);
+    };
+
+
+    template <typename OutputType>
+    void
+    DelayedRejectionMetropolisHastings<OutputType>::
+    sample (const OutputType &starting_point,
+            const std::function<double (const OutputType &)> &log_likelihood,
+            const std::function<std::pair<OutputType,double> (const OutputType &, const std::vector<OutputType>)> &perturb,
+            const unsigned int max_delays,
+            const types::sample_index n_samples)
+    {
+      // Make sure the flush_consumers() function is called at any point
+      // where we exit the current function.
+      Utilities::ScopeExit scope_exit ([this]()
+      {
+        this->flush_consumers();
+      });
+
+      std::mt19937 rng;
+      std::uniform_real_distribution<> uniform_distribution(0,1);
+
+      OutputType current_sample         = starting_point;
+      double     current_log_likelihood = log_likelihood (current_sample);
+
+      // Loop over the desired number of samples
+      for (types::sample_index i=0; i<n_samples; ++i)
+        {
+          // Initialize a vector to store rejected samples
+          std::vector<OutputType> rejected_samples;
+          // Initialize a bool to store whether a sample is accepted
+          bool accepted_sample = false;
+          // Initialize a double to store the previous acceptance_ratio;
+          // this is multiplied to the current acceptance_ratio
+          double prev_acceptance_ratio = 1.0;
+          // Delayed rejection loop
+          for (int delay_stage = 0; delay_stage <= max_delays; ++delay_stage)
+            {
+              // Obtain a new sample by perturbation and evaluate its log likelihood
+              std::pair<OutputType,double> trial_sample_and_ratio = perturb(current_sample, rejected_samples);
+              OutputType trial_sample = std::move(trial_sample_and_ratio.first);
+              const double proposal_distribution_ratio = trial_sample_and_ratio.second;
+              const double trial_log_likelihood = log_likelihood (trial_sample);
+              // Accept trial sample with probability equal to ratio of likelihoods;
+              // (always accept if > 1)
+              double acceptance_ratio = (std::exp(trial_log_likelihood - current_log_likelihood) /
+                                         proposal_distribution_ratio) * prev_acceptance_ratio;
+              if ((trial_log_likelihood - std::log(proposal_distribution_ratio) > current_log_likelihood) ||
+                  acceptance_ratio >= uniform_distribution(rng))
```

Comment:
@bangerth [see this commit](https://github.com/bangerth/SampleFlow/pull/136/commits/11732b93d98a1c432e39cb1e0ebf06beae203727).  I tried with `max_delays=0` just like you said and observed different results from the vanilla MH. I went through the vanilla MH to see what could possibly be different and this was missing. I thought that including this likelihood check was redundant but I guess I was wrong (let's talk about this more next week).  With this included, the `max_delays=0` DRMH produces samples with a mean of `1.99`. 

However, when I increase the number of delay stages to 5, the resulting mean is 1.69. This makes me very suspicous that there is something wrong with the delayed rejection process. What are your thoughts?

Reply:
As for different results: You have to count how many times you evaluate the rng. It produces a reproducible sequence of random numbers, but if you use the rng a different number of times per loop iteration, then at the top of the next iteration you will of course get a different result.